working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+  external factory ThreadPool({required int concurrency});
+  
+  /// Run the given function on this thread pool.
+  /// 
+  /// The [task] is executed in a shared isolate of the current
+  /// isolate group. 
+  ///
+  /// The function will be run in a `Zone` which uses this
+  /// pool as an executor for all callbacks: this
+  /// means the pool will remain alive as long as there is 
+  /// a callback referencing it.
+  void postTask(void Function() task);
+}
+```
+
+We might also want to provide a way to synchronously execute code in a specific isolate on the _current_ thread:
+
+```dart
+class Isolate {
+    T runSync<T>(T Function() cb);
+}
+```
+
+##### Asynchronous code and threads
+
+Consider the following example:
+
+``` dart
+Thread.start(() async {
+  var v = await foo();
+  var u = await bar();
+});
+```
+
+Connecting execution / scheduling behavior to `Zone` allows us to give a clear semantics to this code: this code will run on a specific (newly spawned) thread and will not change threads between suspending and resumptions.  
+
+#### Atomic operations
+
+`Atomic<T>` is a wrapper around a value of type `T` which can be updated atomically. It can only be used with true reference types - an attempt to create an `AtomicRef<int>`, `AtomicRef<double>` , `AtomicRef<(T1, ..., Tn)>` will throw.  
+
+```dart
+// dart:concurrent
+
+class AtomicRef<T extends Shareable> implements Shareable {
+  /// Atomically updates the current value to [desired].
+  ///
+  /// The store has release memory order semantics.
+  void store(T desired);
+
+  /// Atomically reads the current value.
+  ///
+  /// The load has acquire memory order semantics.
+  T load();
+  
+  /// Atomically compares whether the current value is identical to
+  /// [expected] and if it is sets it to [desired] and returns 
+  /// `(true, expected)`.
+  ///
+  /// Otherwise the value is not changed and `(false, currentValue)` is
+  /// returned.
+  (bool, T) compareAndSwap(T expected, T desired);
+}
+
+class AtomicInt32 {
+  void store(int value);
+  int load();
+  (bool, int) compareAndSwap(int expected, int desired);
+
+  int fetchAdd(int v);
+  int fetchSub(int v);
+  int fetchAnd(int v);
+  int fetchOr(int v);
+  int fetchXor(int v);
+}
+
+extension Int32ListAtomics on Int32List {
+  void atomicStore(int index, int value);
+  int atomicLoad(int index);
+  (bool, int) compareAndSwap(int index, int expected, int desired);
+  int fetchAdd(int index, int v);
+  int fetchSub(int index, int v);
+  int fetchAnd(int index, int v);
+  int fetchOr(int index, int v);
+  int fetchXor(int index, int v);  
+}
+
+// These extension methods will only work on fixed-length builtin 
+// List<T> type and will throw an error otherwise.
+extension RefListAtomics<T extends Shareable> on List<T> {
+  void atomicStore(int index, T value);
+  T atomicLoad(int index);
+  (bool, T) compareAndSwap(T expected, T desired);
+}
+
+extension Int32PointerAtomics on Pointer<Int32> {
```

Comment:
Maybe these should live in `dart:ffi` instead of `dart:concurrent`.

Reply:
Tough to say: for now we can just keep them all in one place (for the sake of brevity) and decide later how we want to place them.

Reply:
On the other hand `dart:ffi` changes can be made in isolation - without doing the whole shared memory thing.

Reply:
Split `dart:ffi` into a separate section.